message_ix_models/data/node/R12_SSP_V1.yaml

```diff
@@ -0,0 +1,84 @@
+# Region code list
+#
+# - See message_data.tools.regions.
+# - The ISO 3166-1 alpha-3 codes are not defined in this file, but loaded from
+#   a copy of the ISO database, e.g. in pycountry.
+# - Among others, there are no assignments for:
+#   - ATA Antarctica
+#   - IOT British Indian Ocean Territory
+#   - SGS South Georgia
+
+World:
+  name: World
+  description: R12 regions
+
+R12_AFR:
+  parent: World
+  name: Sub-Saharan Africa
+  description: >-
+    Does not include Saint	Helena,	Ascension	and	Tristan	da	Cunha. 
+    In France-> Mayotte, Reunion data is present post 2010.
+    IEA MAP:  AFRICATOT - (DZA+EGY+LBY+MAR+SDN+SSD+TUN)->IIASA_AFRICA
```

Comment:
It's unclear what this description line and the following line mean, or how they are actually/intended to be used in the code:
- 'AFRICATOT' (for instance) is not defined anywhere.
- 'IIASA_AFRICA' (for instance) is not defined anywhere.
- This added file is not mentioned in doc/pkg-data/node.rst.
- The header comment seems to be exactly the same as the one in R12.yaml.

Reply:
This file was shared by @SiddharthJoshi-Git together with a copy of the preprocessed IEA EWEB 2023. It is used to group and aggregate IEA EWEB data to R12 regions for calibration purposes in MESSAGEix-Materials. Since the file is similar to R12.yaml with main differences in the `child` values, does it make more sense to merge it into the existing R12.yaml? Otherwise we can also rename the file since R12_SSP_V1.yaml is maybe not very indicative.

Reply:
> It is used to group and aggregate IEA EWEB data to R12 regions for calibration purposes in MESSAGEix-Materials.

Can you point to where specifically this happens?

Reply:
All of this happens in [material/data_util.py](https://github.com/iiasa/message-ix-models/blob/7c8cc4c731e8bfec19c1e3a8fca9f7489629fb42/message_ix_models/model/material/data_util.py). The top-level functions are `modify_industry_demand()` and `add_new_ind_hist_act()`, which are called in `model.material.build()`. They read a parquet file containing the EWEB data, map it to R12 and adjust existing `demand` and `historical_activity` parametrization. At the moment there is still some hardcoded parts in the whole procedure and inefficient parts (e.g. reading the parquet file more then once instead of passing it to the respective function).

Reply:
Okay, I see. I recall this was alluded to in a MESSAGE meeting recently—that some code, somewhere, is being used to process the IEA EWEB into a Parquet file, including some spatial manipulations that are not in [`.tools.iea.web`](https://docs.messageix.org/projects/models/en/latest/api/data-sources.html#extended-world-energy-balances-tools-iea-web). Not having seen that code, I can't guess at what it contains, but from the code in `message_ix_models.model.material`/this branch, it seems some of these labels like "AFRICATOT" are introduced or used by it. 

There are several ways in which this appears to duplicate existing `message_ix_models` code that has tests and docs and is more concise. We need to be careful to avoid this. However, to avoid expanding scope too much, I won't insist that they be removed/reworked this PR, but will mention them as further TODOs in #194 or subsidiary to it.

As for what to do in this PR:

Per your link, I see the file is used in two calls to this function https://github.com/iiasa/message-ix-models/blob/36fc249b7991d99b138d371cd8904a5356ac6851/message_ix_models/model/material/data_util.py#L921-L950

I am guessing here because of the lack of tests, but it appears most of the contents of `R12_SSP_V1.yaml` are not used. Rather, only the `child:` field is used to populate a "REGION" column in a data frame that has a "COUNTRY" column. Is that correct?

If so, this can be simplified a lot using existing features. For example, see: https://github.com/iiasa/message-ix-models/blob/dad8f8a5cbdacb85ae8881e267982dad944b8c70/message_ix_models/data/node/R12.yaml#L19
I suspect we can add another ‘annotation’ named, e.g. `material-iea-eweb-region` and containing these values like "CHINAREG". Then the file can be read with [`get_codes()`](https://docs.messageix.org/projects/models/en/latest/api/model.html#message_ix_models.model.structure.get_codes), and [`MappingAdapter`](https://docs.messageix.org/projects/models/en/latest/api/util.html#message_ix_models.util.MappingAdapter) can be used to do the relabeling.

I'd be happy to push some commits to make these changes; although again it will be difficult for me to ensure that I preserve behaviour without tests.

Reply:
Yes I was already wondering how we can harmonize the use of IEA EWEB across the team when I saw the feature being added to `message-ix-models` #25. I agree that this is out of scope for this PR. So we should continue this discussion in a separate issue I guess. The only issue with moving to `tools.iea.web` immediately that I see, is the missing pre-processing that is already applied in the parquet file. We would need to discuss the details with @SiddharthJoshi-Git.

Regarding the use of the `R12_SSP_V1.yaml`: Yes the mapping procedure only uses the `child` field. Adding a iea-eweb field to the original `R12.yaml` sounds like a good idea 👍 Feel free to push the commits that make use of `get_codes()` etc. I will try to write some simple tests today.

Reply:
I've just pushed the modifications, including one added test. Please have a look and try them out with the proprietary data file.

Reply:
Thanks, I manually tested with the private data file now. It seems that parametrization is identical except for R12_EEU. The private data file contains entries, where COUNTRY == "YUG", until 2021. This means there is a double counting of all countries that formerly belonged to Yugoslavia. According to the IEA EWEB documentation, IEA reports Yugoslavia under two codes: YUGOND and MYUGO. But I don't know how YUG data was derived in the private data file. So we need to either revisit the pre-processing of the EWEB data or modify the mapping in R12.yaml.

Reply:
Thanks for doing the check. The annotations I added to R12.yaml include:
```yaml
  material-region: ["*", SCG, YUG, KOSOVO]
```

The file added on the branch has:
```yaml
  child: [ALB, BGR, BIH, CZE, EST, HRV, HUN, LTU, LVA, MKD, MNE, POL, ROU, SRB, SVK, SVN, KOSOVO]
```
Curiously the latter does not include any of YUG, YUGOND, or MYUGO, which really makes me scratch my head.

But anyway, with the goal of identical behaviour, in the former, "*" refers to the list of children for the same node, which I see already includes "YUG". So you could try to edit to:
```yaml
  material-region: ["*", SCG, KOSOVO]
```
(don't duplicate, although with the way the dict is constructed, I don't think this will make any change)

or, most verbose:
```yaml
  material-region: [ALB, BGR, BIH, CZE, EST, HRV, HUN, LTU, LVA, MKD, MNE, POL, ROU, SRB, SVK, SVN, KOSOVO]
```
…which simply copies the value from the added file—without clarifying why the Yugoslavia-related codes are omitted.

Could you please try those two settings and see what happens?

Reply:
Applying the second suggested setting leads to the correct parametrization. So this would be a viable solution.

Reply:
Okay, good to confirm. If this works, then perhaps you can `git rebase -i main` your branch and put `d`/`drop` beside the commit that added the R12_SSP_V1.yaml file (if it was a single commit. If it was added with other files, try `e`/`edit` and adjust the commit so the file is not added).

It may also help to:
- Put a comment in R12.yaml file directly above that line, linking back to this thread or with a summary, in case others are also confused.
- Expand the header comment in R12.yaml to inform people where they could find the R12_SSP_V1.yaml from which we got that information.